kernel/declarations.ml

```diff
@@ -267,6 +267,18 @@ type module_alg_expr =
   | MEapply of module_alg_expr * ModPath.t
   | MEwith of module_alg_expr * with_declaration
 
+type ('body, 'expr, _) module_implementation =
+  | Abstract : ('body, 'expr, [ `modimpl ]) module_implementation
+    (** no accessible implementation *)
+  | Algebraic : 'expr ->  ('body, 'expr, [ `modimpl ]) module_implementation
+    (** non-interactive algebraic expression *)
+  | Struct : 'body ->  ('body, 'expr, [ `modimpl ]) module_implementation
+    (** interactive body *)
+  | FullStruct : ('body, 'expr, [ `modimpl ]) module_implementation
+    (** special case of [Struct] : the body is exactly [mod_type] *)
+  | ModType : ('body, 'expr, [ `modtype ]) module_implementation
```

Comment:
That seems a bit weird, how is a modtype a module implementation? Shouldn't we got into the opposite direction?

Also, IMHO the use of variants here is overkill.

Reply:
We're already reusing a lot of the same data structures for both modules and module types.

> Also, IMHO the use of variants here is overkill.

It constrains the fact that there are only two possible cases, which was not enforced by the previous state. (It does simplify a bit the functions having to take HO arguments for substitution and hashconsing for instance.)